File: js/타겟 넘버(dfs, bfs).js
// bfs
function solution(numbers, target) {
    var answer = 0;
    var leaves = [0];
    
    for (let number of numbers) {
        var tmpArr = [];
        for (let leaf of leaves) {
            tmpArr.push(leaf + number);
            tmpArr.push(leaf - number);
        }
        leaves = tmpArr;
    }
    var answer = leaves.filter(e => target === e).length;
    return answer;
}

// dfs
var answer = 0;

function dfs(numbers, target, current, index) {
    if (numbers.length === index) {
        if (current === target) {
            answer++;
        }
        return
    }
    
    dfs(numbers, target, current + numbers[index], index + 1);
    dfs(numbers, target, current - numbers[index], index + 1);
}

function solution(numbers, target) {
    dfs(numbers, target, 0, 0);
    return answer;
}

File: python/타겟 넘버(dfs, bfs).py
# bfs
def solution(numbers, target):
    leaves = [0]
    for number in numbers:
        tmp = []
        for leaf in leaves:
            tmp.append(leaf + number)
            tmp.append(leaf - number)
        leaves = tmp
    answer = leaves.count(target)
    return answer

# dfs
answer = 0

def dfs(numbers, target, current, index):
    global answer
    if len(numbers) == index:
        if current == target:
            answer += 1
        return
    
    dfs(numbers, target, current + numbers[index], index + 1)
    dfs(numbers, target, current - numbers[index], index + 1)


def solution(numbers, target):
    dfs(numbers, target, 0, 0)
    return answer 